File: app/app/components/utils/akave-api.js
import axios from "axios";

const API_BASE_URL = "https://cb62-102-91-104-117.ngrok-free.app";

async function apiRequest(method, endpoint, data = null) {
  try {
    const response = await axios({
      method,
      url: `${API_BASE_URL}${endpoint}`,
      data,
    });
    return response.data;
  } catch (error) {
    console.error(error.response ? error.response.data : error.message);
    throw error;
  }
}

export async function createBucket(bucketName) {
  return apiRequest("POST", "/buckets", { bucketName });
}

export async function uploadFile(bucketName, fileName, fileContent) {
  await ensureBucket(bucketName);
  const form = new FormData();
  const blob = new Blob([JSON.stringify(fileContent)], {
    type: "application/json",
  });
  form.append("file", blob, fileName);

  const response = await axios.post(
    `${API_BASE_URL}/buckets/${bucketName}/files`,
    form
  );
  return response.data; // { success: true, data: { RootCID: "..." } }
}

export async function downloadFile(bucketName, fileName) {
  const response = await axios.get(
    `${API_BASE_URL}/buckets/${bucketName}/files/${fileName}/download`,
    {
      responseType: "json",
    }
  );
  return response.data;
}

export async function listFiles(bucketName) {
  const response = await axios.get(
    `${API_BASE_URL}/buckets/${bucketName}/files`
  );
  return response.data.data; // Array of {Name, RootCID, ...}
}

export async function ensureBucket(bucketName) {
  try {
    const response = await axios.get(`${API_BASE_URL}/buckets`);
    const buckets = response.data.data || [];
    const bucketExists = buckets.some((b) => b.Name === bucketName);
    if (bucketExists) {
      console.log(`Bucket ${bucketName} already exists`);
      return;
    }
    await axios.post(`${API_BASE_URL}/buckets`, { bucketName });
    console.log(`Bucket ${bucketName} created`);
  } catch (error) {
    console.error(
      "Failed to ensure bucket:",
      error.response?.data || error.message
    );
    throw error; // Let the app handle it, but it won’t hit this now
  }
}

File: backend-ml/akave_api.py
import requests
import json

API_BASE_URL = "https://c57a-102-91-105-198.ngrok-free.app"

def listFiles(bucket_name):
    url = f"{API_BASE_URL}/buckets/{bucket_name}/files"
    response = requests.get(url)
    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(f"Failed to list files: {response.text}")

def downloadFile(bucket_name, file_name):
    url = f"{API_BASE_URL}/buckets/{bucket_name}/files/{file_name}/download"
    response = requests.get(url)
    if response.status_code == 200:
        try:
            return response.json()  # for patient JSON files
        except ValueError:
            return response.content  # for binary files like .h5
    else:
        raise Exception(f"Failed to download file: {response.text}")

def download_model(bucket_name, file_name, output_path):
    content = downloadFile(bucket_name, file_name)
    with open(output_path, "wb") as f:
        f.write(content)
    print(f"Downloaded {file_name} to {output_path}")


# python3 -m venv venv
# source venv/bin/activate
